panon/backend/source.py

```diff
@@ -103,7 +103,11 @@ def read(self):
         return data
 
     def start(self):
-        from soundcard import pulseaudio as sc
+        import soundcard as sc
```

Comment:
I doubt this will make all this work on windows, but I think this is the intended way of importing soundcard.

Reply:
> I doubt this will make all this work on windows and macOS

lol

> but I think this is the intended way of importing soundcard.

I decided to package soundcard into `panon.plasmoid`, because soundcard is not available in ArchLinux's official repo, guess neither available in Ubuntu's repo. And I also want to keep the 
plasmoid package as small as possible, [so I removed every files from soundcard other than pulseaudio.py and pulseaudio.py.h.](https://github.com/rbn42/panon/tree/master/plasmoid/contents/scripts/soundcard) That's why pulseaudio submodule is imported directly here. 

So, do you still think it is better to import soundcard's main module? Or maybe we should remove soundcard, and ask KDE store users to install soundcard by pip? 

BTW, if you agree to retain soundcard files, can you also upgrade [third_party/SoundCard](https://github.com/rbn42/panon/tree/master/third_party) in this PR?

Reply:
The AUR package for panon actually doesn’t use your bundled SoundCard version, instead using the AUR version.

Standard practice on Arch is that when a AUR package is moved to the official packages, all dependencies are too.

I personally don’t like bundling 3rd party stuff and instead rely on package managers to handle that, unless there’s a really good reason to do it, like having to patch the 3rd party thing.

Reply:
> Standard practice on Arch is that when a AUR package is moved to the official packages, all dependencies are too.

I mentioned Arch and Ubuntu because I think they are the most popular distros. But panon is provided as a KDE Store package, which does not benefit from the package managers, pacman or pip. I believe the majority install panon through KDE Store, not AUR or github. Then, is it still necessary to apply standard practice on Arch to a KDE Store package?

And panon on AUR is an orphan. And I don't have knowledge of AUR conventions, so I don't want to adopt panon on AUR.

> I personally don’t like bundling 3rd party stuff and instead rely on package managers to handle that,

Yes, bundling 3rd party stuff, like the way of Flatpak or Snap. Since KDE Store does not handle dependencies like pacman. I thought bundling stuff should be permitted. 

But if you insist, I will go with your choice. Since I don't have much experience on choices like this. That means removing 3rd party stuff from this repo, and asking KDE Store users to install soundcard through pip, right?

Reply:
And, soundcard depends on `python-cffi`. I am not familiar with it. But if this package is not installed, I guess pip will compile this package when people try to install soundcard, which makes the installation even more complex.

So, we still have to tell KDE Store users to install soundcard's dependency `python-cffi` through the package manager of their distro, instead of pip.

Reply:
> panon on AUR is an orphan

I can take it, no problem, I have a lot of AUR packages

Other than that, I think *ideally*, the store package should contain as many bundled deps as possible while a .deb or pacman/AUR package should depend on system dependencies.

We could create a script that downloads and bundles soundcard and a compiled version of cffi together with your code and creates a store package from it, right?

Reply:
> I can take it, no problem, I have a lot of AUR packages

That would be nice. Thank you. 

> We could create a script that downloads and bundles soundcard and a compiled version of cffi together with your code and creates a store package from it, right?

Right. Basically this is what I did.
1. I added soundcard as a git submodule. So soundcard was expected to be downloaded by the command `git submodule update --init`.
2. And here is [a script.](https://github.com/rbn42/panon/blob/master/makepackage.sh) to create the store package. The script packages the [plasmoid/](https://github.com/rbn42/panon/tree/master/plasmoid) folder into a store package.
3. There are [soft links](https://github.com/rbn42/panon/tree/master/plasmoid/contents/scripts/soundcard) to soundcard files in the [plasmoid/](https://github.com/rbn42/panon/tree/master/plasmoid) folder. So these files will be added to the store package by the script. This is my way to bundle soundcard files. 
4. I added only necessary soundcard files to keep the store package as small as possible. That is why `from soundcard import pulseaudio as sc` here. Because soundcard's main module does not exist in the store package.

OK, so if you want to use `import soundcard as sc` here, we need to add the entire soundcard packages back to the store package. Well, you can do it this PR or leave it to me.

Reply:
>  the entire soundcard packages
Wrong. Maybe only __init__.py is required.

Reply:
BTW, if you agree to retain soundcard as a git submodule, could you please also upgrade third_party/SoundCard in this PR?

Reply:
> There are soft links to soundcard files in the plasmoid/ folder. So these files will be added to the store package by the script. This is my way to bundle soundcard files.

got it. creative!

> Maybe only `__init__.py` is required.

exactly, because the only thing it does in our case is `if sys.platform == 'linux': from .pulseaudio import *`.

---

so I’ll create the `__init__.py` symlink, update the submodule, and we’re good to go.